tests/test_models.py

```diff
@@ -0,0 +1,1194 @@
+    prepare_charmcraft_yaml(
+        dedent(
+            """
+            type: charm
+            bases:
+              - name: test-name
+                channel: test-channel
+            """
+        ),
+    )
+    prepare_metadata_yaml(
+        dedent(
+            """
+            summary: test-summary
+            description: test-description
+            """
+        ),
+    )
+
+    with pytest.raises(CraftError, match="field 'name' required in top-level configuration"):
+        load(tmp_path)
+
+
+def test_load_minimal_metadata_from_metadata_yaml_missing_type(
+    tmp_path, prepare_charmcraft_yaml, prepare_metadata_yaml
+):
+    """Load a mimimal charmcraft.yaml with metadata.yaml. But missing type."""
+    prepare_charmcraft_yaml(
+        dedent(
+            """
+            bases:
+              - name: test-name
+                channel: test-channel
+            """
+        ),
+    )
+    prepare_metadata_yaml(
+        dedent(
+            """
+            name: test-charm-name
+            summary: test-summary
+            description: test-description
+            """
+        ),
+    )
+
+    with pytest.raises(CraftError, match="field 'type' required in top-level configuration"):
+        load(tmp_path)
+
+
+def test_load_minimal_metadata_from_metadata_yaml_missing_summary(
+    tmp_path, prepare_charmcraft_yaml, prepare_metadata_yaml
+):
+    """Load a mimimal charmcraft.yaml with metadata.yaml. But missing summary."""
+    prepare_charmcraft_yaml(
+        dedent(
+            """
+            type: charm
+
+            bases:
+                - name: test-name
+                  channel: test-channel
+            """
+        ),
+    )
+    prepare_metadata_yaml(
+        dedent(
+            """
+            name: test-charm-name
+            description: test-description
+            """
+        ),
+    )
+
+    with pytest.raises(CraftError, match="field 'summary' required in top-level configuration"):
+        load(tmp_path)
+
+
+def test_load_minimal_metadata_from_metadata_yaml_missing_description(
+    tmp_path, prepare_charmcraft_yaml, prepare_metadata_yaml
+):
+    """Load a mimimal charmcraft.yaml with metadata.yaml. But missing description."""
+    prepare_charmcraft_yaml(
+        dedent(
+            """
+            type: charm
+
+            bases:
+              - name: test-name
+                channel: test-channel
+            """
+        ),
+    )
+    prepare_metadata_yaml(
+        dedent(
+            """
+            name: test-charm-name
+            summary: test-summary
+            """
+        ),
+    )
+
+    with pytest.raises(
+        CraftError, match="field 'description' required in top-level configuration"
+    ):
+        load(tmp_path)
+
+
+def test_load_full_metadata_from_charmcraft_yaml(tmp_path, prepare_charmcraft_yaml):
+    """Load a charmcraft.yaml with full metadata. (Spec ST087)"""
+    prepare_charmcraft_yaml(
+        dedent(
+            """
```

Comment:
@dboddie this might be worth bookmarking for when we tackle improvements to charmcraft's docs. This (and its companion at ST087) might be the most complete description of charms that we have.